### PR TITLE
Add a new, general purpose convolution function

### PR DESCRIPTION
**Pull request by tskisner**
In several places throughout the code we perform Fourier domain convolution and use a variety of techniques.  This work aims to have a common function for some of these use cases which carefully handles the symmetric extension of data in the Fourier domain buffer and is unit tested.  Changes include:

- New toast.fft.convolve function that can take precomputed kernels or use a callback function to generate the kernels.  Unit tests verify roundtrip and expected sample phase shift introduced by a Butterworth filter.

- Porting of the timeconstant deconvolution operator to this new function.

- Small unrelated fix to toast timing plots

- Unrelated fixes to DetectorData indexing.  In a couple places we were triggering data copying by using numpy "advanced indexing".

### CODE REVIEW

**Comment by tskisner** [comment]
Some diagnostic plots from the unit tests:

Input fake data with 2 sine waves:
conv_common_in_0-500 pdf
After applying a butterworth low-pass using both numpy and toast internal fft:
conv_common_compare_output_0-500 pdf
Difference between those two code paths:
conv_common_compare_diff_0-500 pdf
Comparison with the input low-frequency sine wave after phase shifting by the expected amount considering the filter kernel:
conv_common_compare_shifted_0-500 pdf

File: src/toast/ops/time_constant.py [hunk 1]
```diff
@@ -51,12 +59,23 @@ class TimeConstant(Operator):
         help="Detector flag mask for cutting detectors with invalid Tau values.",
     )
 
+    edge_flag_mask = Int(
```
Comment (keskitalo) [inline src/toast/ops/time_constant.py:62]: Excellent, this was definitely missing.

**Review by keskitalo — APPROVED** [review]
Looks great!

File: src/toast/fft.py [hunk 2]
```diff
@@ -101,3 +103,314 @@ def r1d_backward(indata):
...
+            kre = np.real(krn[1:-1])
+            kim = np.imag(krn[1:-1])
+            fre = np.array(fd[1 : n_psd - 1])
+            fim = np.array(fd[-1 : n_psd - 1 : -1])
+            nyq = fd[n_psd - 1]
+            # We handle the zero and nyquist frequencies separately
+            if deconvolve:
+                denom = kre**2 + kim**2
+                # Real values
+                fd[1 : n_psd - 1] = (fre * kre + fim * kim) / denom
```
[the remark this answers is not in the document]
Comment (tskisner) [reply src/toast/fft.py:378]: ... and same resolution.  The complex magnitude of small kernel values is set to minimum limit.

File: src/toast/fft.py [hunk 3]
```diff
@@ -101,3 +103,314 @@ def r1d_backward(indata):
         for indx in range(count):
             ret[indx, :] = plan.tdata(indx)
         return ret
+
+
+def convolve(
```
[the remark this answers is not in the document]
Comment (tskisner) [reply src/toast/fft.py:108]: This is a fair point.  The internal helper functions make use of previously-defined variables internal to the `convolve()` function.  So moving the internal functions outside would require passing a bunch of other variables for no other reason than the fact that they are no longer internal functions.

I am going to leave this function as-is for now.  It is ~300 lines long, which is getting close to the limit of easy readability.  If we need to extend this function in the future, then we should pull out those internal functions and add the boilerplate needed to pass in the additional parameters.

**Pull request merged**